qiskit_experiments/calibration/calibrations.py

```diff
@@ -0,0 +1,647 @@
+        """
+        Add a schedule and register its parameters.
+
+        Args:
+            schedule: The schedule to add.
+            qubits: The qubits for which to add the schedules. If None is given then this
+                schedule is the default schedule for all qubits.
+
+        Raises:
+            CalibrationError: If the parameterized channel index is not formatted
+                following index1.index2... or if several parameters in the same schedule
+                have the same name.
+        """
+        # check that channels, if parameterized, have the proper name format.
+        # pylint: disable = raise-missing-from
+        param_indices = set()
+        for ch in schedule.channels:
+            if isinstance(ch.index, Parameter):
+                param_indices.add(ch.index)
+                if re.compile(self._channel_pattern).match(ch.index.name) is None:
+                    raise CalibrationError(
+                        f"Parameterized channel must correspond to {self._channel_pattern}"
+                    )
+
+        self._schedules[(schedule.name, qubits)] = schedule
+
+        param_names = [param.name for param in schedule.parameters]
+
+        if len(param_names) != len(set(param_names)):
+            raise CalibrationError(f"Parameter names in {schedule.name} must be unique.")
+
+        # Register parameters that are not indices.
+        for param in schedule.parameters:
+            if param not in param_indices:
+                self.register_parameter(param, schedule, qubits)
+
+    def register_parameter(
+        self, parameter: Parameter, schedule: Schedule = None, qubits: Tuple = None
+    ):
+        """
+        Registers a parameter for the given schedule. This allows self to determine the
+        parameter instance that corresponds to the given schedule name, parameter name
+        and qubits.
+
+        Args:
+            parameter: The parameter to register.
+            schedule: The Schedule to which this parameter belongs. The schedule can
+                be None which allows the calibration to accommodate, e.g. qubit frequencies.
+            qubits: The qubits for which to register the parameter.
+        """
+        sched_name = schedule.name if schedule else None
+        key = ParameterKey(sched_name, parameter.name, qubits)
+        self._parameter_map[key] = parameter
+        self._parameter_map_r[parameter].add(key)
+
+    @property
+    def parameters(self) -> Dict[Parameter, Set[ParameterKey]]:
+        """
+        Returns a dictionary mapping parameters managed by the calibrations to the schedules and
+        qubits and parameter names using the parameters. The values of the dict are sets containing
+        the parameter keys. Parameters that are not attached to a schedule will have None in place
+        of a schedule name.
+        """
+        return self._parameter_map_r
+
+    def add_parameter_value(
+        self,
+        value: ParameterValue,
```

Comment:
Can this be `Union[ParameterValue, float]` by assigning timestamp at the time this method is called? `ParameterValue` seems to be an internal data container, and end-uses should be able to use this class without learning this.

Reply:
I agree. This makes the framework more flexible. See here: b314294. I think in most cases the calibration experiment will take care of generating an instance of `ParameterValue` but I think it does not hurt to have more flexibility.